src/bridge/cockpitpackages.c

```diff
@@ -0,0 +1,1040 @@
+
+  dir = g_dir_open (directory, 0, &error);
+  if (error != NULL)
+    {
+      g_warning ("couldn't list directory: %s: %s", directory, error->message);
+      g_error_free (error);
+      return FALSE;
+    }
+
+  names = g_ptr_array_new ();
+  for (;;)
+    {
+      name = g_dir_read_name (dir);
+      if (!name)
+        break;
+      g_ptr_array_add (names, g_strdup (name));
+    }
+
+  g_dir_close (dir);
+
+  g_ptr_array_sort (names, compare_filenames);
+  g_ptr_array_add (names, NULL);
+
+  return (gchar **)g_ptr_array_free (names, FALSE);
+}
+
+static gboolean
+package_checksum_directory (GChecksum *checksum,
+                            const gchar *root,
+                            const gchar *directory)
+{
+  gboolean ret = FALSE;
+  gchar *path = NULL;
+  gchar **names = NULL;
+  gchar *filename;
+  gint i;
+
+  path = g_build_filename (root, directory, NULL);
+  names = directory_filenames (path);
+  if (!names)
+    goto out;
+
+  ret = TRUE;
+  for (i = 0; names[i] != NULL; i++)
+    {
+      if (directory)
+        filename = g_build_filename (directory, names[i], NULL);
+      else
+        filename = g_strdup (names[i]);
+      ret = package_checksum_file (checksum, root, filename);
+      g_free (filename);
+      if (!ret)
+        goto out;
+    }
+
+out:
+  g_free (path);
+  g_strfreev (names);
+  return ret;
+}
+
+static JsonObject *
+read_package_manifest (const gchar *directory,
+                       const gchar *package)
+{
+  JsonObject *manifest = NULL;
+  GError *error = NULL;
+  GMappedFile *mapped;
+  gchar *filename;
+  GBytes *bytes;
+
+  filename = g_build_filename (directory, "manifest.json", NULL);
+  mapped = g_mapped_file_new (filename, FALSE, &error);
+  if (!mapped)
+    {
+      if (g_error_matches (error, G_FILE_ERROR, G_FILE_ERROR_NOENT))
+        g_debug ("no manifest found: %s", filename);
+      else if (!g_error_matches (error, G_FILE_ERROR, G_FILE_ERROR_NOTDIR))
+        g_message ("%s: %s", package, error->message);
+      g_clear_error (&error);
+    }
+  else
+    {
+     if (!validate_package (package))
+       {
+         g_warning ("package has invalid name: %s", package);
+       }
+     else
+       {
+         bytes = g_mapped_file_get_bytes (mapped);
+         manifest = cockpit_json_parse_bytes (bytes, &error);
+         g_bytes_unref (bytes);
+
+         if (!manifest)
+           {
+             g_message ("%s: invalid manifest: %s", package, error->message);
+             g_clear_error (&error);
+           }
+       }
+
+      g_mapped_file_unref (mapped);
+    }
+
+  g_free (filename);
+  return manifest;
+}
+
+static CockpitPackage *
+maybe_add_package (GHashTable *listing,
+                   const gchar *parent,
+                   const gchar *name,
+                   GChecksum *checksum)
+{
+  CockpitPackage *package = NULL;
+  gchar *path = NULL;
+  JsonObject *manifest = NULL;
+
+  package = g_hash_table_lookup (listing, name);
+  if (package)
+    goto out;
```

Comment:
(And this is the reason why my old trick of symlinking /usr/share/cockpit into ~/.local/share doesn't switch off minification anymore.)